src/Unit/Algebra.hs

```diff
@@ -0,0 +1,125 @@
+{-# LANGUAGE AllowAmbiguousTypes #-}
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE DeriveTraversable #-}
+{-# LANGUAGE DerivingVia #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE KindSignatures #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE TypeOperators #-}
+{-# LANGUAGE UndecidableInstances #-}
+module Unit.Algebra
+(  -- * Algebra
+  Mul(..)
+, (./.)
+  -- * Combinators
+, (:*:)(..)
+, Inv(..)
+, (:/:)
+) where
+
+import Data.Functor.Const
+import Data.Functor.Identity
+import Foreign.Storable
+import GL.Type as GL
+import GL.Uniform
+import Linear.Conjugate
+import Linear.Epsilon
+import Linear.Metric
+import Linear.Vector
+import Unit
+
+-- * Algebra
+
+class (Unit u, Unit v) => Mul u v where
+  type Res u v :: * -> *
+  (.*.) :: Fractional a => u a -> v a -> Res u v a
```

Comment:
I might tear this out and go with `units` at some point, but for now this will do.